### PR TITLE
Add Cloud Armor in front of this deployed demo

### PR DESCRIPTION
**Pull request by mathieu-benoit**
Document the way to setup Cloud Armor in front of the public endpoint of this demo onlineboutique.dev.

_Note: not yet actually deployed, opening this PR as a proposition to discuss before moving forward if making sense._

What's in this PR:
- Document the steps do deploy the `release-cluster` part with more automation with `gcloud`
- Add `BackendConfig` to bind Cloud Armor to the `Ingress`
- Add `FrontendConfig` to bind the SSL Policy and redirect http to https
- Update `Ingress`'s `apiVersion` from `apiVersion: networking.k8s.io/v1beta1` to `networking.k8s.io/v1` to avoid future breaking change because [deprecation with GKE 1.22](https://cloud.google.com/kubernetes-engine/docs/deprecations/apis-1-22#ingress-v122)
- Remove the `frontend-nodeport.yaml` because it's not anymore a requirement, the default `ClusterIP` one is sufficient/working now.
- To avoid to have another Public IP, do a `kubectl delete svc frontend-external` in the process too
- Not related, fix a typo in the Memorystore (redis) doc.

Next steps:
- [X] Test on a personal GKE cluster (see comment below with details information about the tests)
- [x] Validate with the team about their thoughts and feedback
- [ ] If validated, apply this to onlineboutique.dev
- [ ] Eventually, port this to BoA too, to be consistent

### CODE REVIEW

**Comment by mathieu-benoit** [comment]
Tests conducted:
<details>
<summary>Bash commands to provision GKE/ASM and test this PR...</summary>
<p>

```bash
PROJECT_ID=FIXME
gcloud config set project $PROJECT_ID
k create ns onlineboutique
k apply -f release/kubernetes-manifests.yaml -n onlineboutique
gcloud compute addresses create online-boutique-ip --global
PUBLIC_IP=$(gcloud compute addresses describe online-boutique-ip \
    --global \
    --format "value(address)")
   
export HOST_NAME="onlineboutique.endpoints.${PROJECT_ID}.cloud.goog"
cat <<EOF > dns-spec.yaml
swagger: "2.0"
info:
  description: "Cloud Endpoints DNS"
  title: "Cloud Endpoints DNS"
  version: "1.0.0"
paths: {}
host: "${HOST_NAME}"
x-google-endpoints:
- name: "${HOST_NAME}"
  target: "${PUBLIC_IP}"
EOF
gcloud endpoints services deploy dns-spec.yaml
SECURITY_POLICY_NAME=online-boutique-security-policy
gcloud compute security-policies create $SECURITY_POLICY_NAME \
    --description "Block XSS attacks"
gcloud compute security-policies rules create 1000 \
    --security-policy $SECURITY_POLICY_NAME \
    --expression "evaluatePreconfiguredExpr('xss-stable')" \
    --action "deny-403" \
    --description "XSS attack filtering"
gcloud compute security-policies rules create 12345 \
    --security-policy $SECURITY_POLICY_NAME \
    --expression "evaluatePreconfiguredExpr('cve-canary')" \
    --action "deny-403" \
    --description "CVE-2021-44228 and CVE-2021-45046"
gcloud compute security-policies update $SECURITY_POLICY_NAME \
    --enable-layer7-ddos-defense
gcloud compute security-policies update $SECURITY_POLICY_NAME \
    --log-level=VERBOSE
SSL_POLICY_NAME=online-boutique-ssl-policy
gcloud compute ssl-policies create $SSL_POLICY_NAME \
    --profile COMPATIBLE  \
    --min-tls-version 1.0
sed -i "s,onlineboutique.dev,${HOST_NAME},g" .github/release-cluster/managed-cert.yaml
kubectl apply -f .github/release-cluster/ -n onlineboutique
```

</p>
</details>

**Comment by mathieu-benoit** [comment]
Ready for your review and comments, thanks.

**Comment by NimJay** [comment]
Thanks so much for creating this pull-request. And apologies for the late review.
I think using Cloud Armor is a good idea — we would be dog-fooding yet another GCP product! :)

Warning: I have added a commit to the `mathieu-benoit/cloud-armor-release-cluster` branch, with some wording changes. 
So you may want to `git pull` from your end.


**Question:** 
See [hack/README.md](https://github.com/GoogleCloudPlatform/microservices-demo/tree/main/hack#9-connect-to-our-online-boutique-release-gke-cluster) where we've documented the release process.
Step 10 consists of applying `./release/kubernetes-manifests.yaml` to our deployment GKE cluster.
Do you think we should add a step between step 10 and step 11 similar to:
> 11. Delete unused resources.
> ```
> kubectl delete service frontend-external
> kubectl delete deployment loadgenerator
> ```

**Comment by NimJay** [comment]
Ah, that's a good idea. 
I like the idea of not duplicating instructions.
Got for it. :)

**Review by NimJay — APPROVED** [review]
Just finished testing this on my own cluster and domain.
See [deletemenow.peachytools.com](https://deletemenow.peachytools.com) (which I'll take down once this PR is merged).
Cloud Armor was applied:
Screen Shot 2022-01-27 at 2 00 07 PM

I didn't test Cloud Armor itself — but I trust that it works.
I left 2 comments.
They're minor so I'll approve this.

Thanks again for doing this, @mathieu-benoit! And thanks for the very details PR description. It made the changes easy to review. 👏

I'll apply the changes to [onlineboutique.dev](https://onlineboutique.dev/) as soon as this is merge. :)

File: .github/release-cluster/README.md [hunk 1]
```diff
@@ -1,3 +1,65 @@
...
+
+Set up an SSL policy in order to later set up a redirect from HTTP to HTTPs:
+```
+SSL_POLICY_NAME=online-boutique-ssl-policy # Name hard-coded in: frontendconfig.yaml
+gcloud compute ssl-policies create $SSL_POLICY_NAME \
+    --profile COMPATIBLE  \
+    --min-tls-version 1.0
+```
+
+Deploy the Kubernetes manifests:
```
[the remark this answers is not in the document]
Comment (mathieu-benoit) [reply .github/release-cluster/README.md]: Agreed, done.

File: .github/release-cluster/README.md [hunk 2]
```diff
@@ -1,3 +1,65 @@
...
+gcloud compute addresses describe $STATIC_IP_NAME \
+    --global \
+    --format "value(address)"
+```
+
+Set up Cloud Armor:
+```
+SECURITY_POLICY_NAME=online-boutique-security-policy # Name hard-coded in: backendconfig.yaml
+gcloud compute security-policies create $SECURITY_POLICY_NAME \
+    --description "Block XSS attacks"
```
[the remark this answers is not in the document]
Comment (mathieu-benoit) [reply .github/release-cluster/README.md]: Good catch, done.

**Comment by mathieu-benoit** [comment]
Thanks for the review and comments @NimJay, let me know if there is anything else.

**Comment by NimJay** [comment]
I just applied the changes to the onlineboutique.dev cluster ([online-boutique-release](https://pantheon.corp.google.com/kubernetes/list/overview?referrer=search&project=online-boutique-ci)).
For some reason, the [`online-boutique-security-policy`](https://pantheon.corp.google.com/net-security/securitypolicies/list?referrer=search&project=online-boutique-ci) has "0 targets".
I'll give it some time and check back in a bit.

**Comment by NimJay** [comment]
## Update on Investigation 

@mathieu-benoit and I just investigated a bit. 

The `frontend-ingress` is still connected to the old `frontend-nodeport`:
```
kubectl describe ingress
Name:             frontend-ingress
Labels:           <none>
Namespace:        default
Address:          35.244.250.164
Default backend:  frontend-nodeport:80 (<error: endpoints "frontend-nodeport" not found>)
```
which I had deleted using: 
```
kubectl delete Service frontend-nodeport
```
after I applied the new manifests from the `./github/release-cluster/` folder.

But we just reapplied the old `frontend-nodeport.yaml` file.
And now we see that the Cloud Armor [`online-boutique-security-policy`](https://pantheon.corp.google.com/net-security/securitypolicies/list?referrer=search&project=online-boutique-ci) has "1 target".

**Comment by mathieu-benoit** [comment]
Found out the issue: because of the update of the ingress to v1, I haven't set the defaultBakend field. The issue was not there on a brand new GKE cluster we tested, but because the Ingress of the live cluster of onlineboutique.dev was pointing to the deleted frontendservice-nodeport, it raised that issue.

All good now, #698 is fixing the issue.

**Pull request merged**